src/ecmult_impl.h

```diff
@@ -718,31 +820,51 @@ static int secp256k1_ecmult_pippenger_batch_single(const secp256k1_callback* err
- * a given scratch space. The function ensures that fewer points may also be
- * used.
+/* Returns the (near) maximum number of points in addition to G that can be
+ * used with a given scratch space. It may not return the actual maximum number
+ * of points possible. Otherwise, fewer points would not fit into the scratch
+ * space in general. If a scratch space has exactly
+ * `pippenger_scratch_size(n_points)` left, then
+ * `pippenger_max_points(cb, scratch) <= n_points`.
  */
 static size_t secp256k1_pippenger_max_points(const secp256k1_callback* error_callback, secp256k1_scratch *scratch) {
-    size_t max_alloc = secp256k1_scratch_max_allocation(error_callback, scratch, PIPPENGER_SCRATCH_OBJECTS);
+    /* Call max_allocation with 0 objects because otherwise it would assume
+     * worst case padding but in this function we want to be exact. */
+    size_t max_alloc = secp256k1_scratch_max_allocation(error_callback, scratch, 0);
     int bucket_window;
     size_t res = 0;
 
     for (bucket_window = 1; bucket_window <= PIPPENGER_MAX_BUCKET_WINDOW; bucket_window++) {
         size_t n_points;
         size_t max_points = secp256k1_pippenger_bucket_window_inv(bucket_window);
         size_t space_for_points;
-        size_t space_overhead;
-        size_t entry_size = sizeof(secp256k1_ge) + sizeof(secp256k1_scalar) + sizeof(struct secp256k1_pippenger_point_state) + (WNAF_SIZE(bucket_window+1)+1)*sizeof(int);
+        size_t space_constant;
+        /* Compute entry size without taking alignment into account */
+        size_t entry_size = secp256k1_pippenger_scratch_size_points(0, bucket_window, 0);
 
-        entry_size = 2*entry_size;
-        space_overhead = (sizeof(secp256k1_gej) << bucket_window) + entry_size + sizeof(struct secp256k1_pippenger_state);
-        if (space_overhead > max_alloc) {
+        space_constant = secp256k1_pippenger_scratch_size_constant(bucket_window);
+        if (space_constant + entry_size > max_alloc) {
             break;
         }
-        space_for_points = max_alloc - space_overhead;
+        space_for_points = max_alloc - space_constant;
+
+        /* Compute an upper bound for the number excluding the base point G.
+         * It's an upper bound because alignment is not taken into account. */
+        n_points = space_for_points / entry_size - 1;
+        /* Compute an upper bound for the number of points after subtracting
+         * space for the base point G. It's an upper bound because alignment is
+         * not taken into account. */
+        n_points = (space_for_points - entry_size)/entry_size;
+        if (n_points > 0
+              && space_for_points < secp256k1_pippenger_scratch_size_points(n_points, bucket_window, 1)) {
+            /* If there's not enough space after alignment is taken into
+             * account, it suffices to decrease n_points by one. This is because
+             * the maximum padding required is less than an entry. */
+            n_points -= 1;
```

Comment:
Do you know how it's possible to hit this branch in a test (I wasn't able to do so even by trying a huge number of scratch sizes)?